waku/v2/utils/pagination.nim

```diff
@@ -14,26 +14,52 @@ export hash
 type
   Index* = object
     ## This type contains the  description of an Index used in the pagination of WakuMessages
-    digest*: MDigest[256]
+    digest*: MDigest[256] # calculated over payload and content topic
     receiverTime*: Timestamp
     senderTime*: Timestamp # the time at which the message is generated
+    pubsubTopic*: string
 
 proc `==`*(x, y: Index): bool =
-  ## receiverTime plays no role in index comparison
-  (x.senderTime == y.senderTime) and (x.digest == y.digest)
+  ## receiverTime plays no role in index equality
+  (x.senderTime == y.senderTime) and
+  (x.digest == y.digest) and
+  (x.pubsubTopic == y.pubsubTopic)
 
 proc cmp*(x, y: Index): int =
   ## compares x and y
   ## returns 0 if they are equal 
   ## returns -1 if x < y
   ## returns 1 if x > y
-  ## receiverTime plays no role in index comparison
+  ## 
+  ## Default sorting order priority is:
+  ## 1. senderTimestamp
+  ## 2. receiverTimestamp (a fallback only if senderTimestamp unset and all other fields unequal)
+  ## 3. message digest
+  ## 4. pubsubTopic
   
   # Timestamp has a higher priority for comparison
   let timecmp = cmp(x.senderTime, y.senderTime)
   if timecmp != 0: 
     return timecmp
 
-  # Only when timestamps are equal 
-  let digestcm = cmp(x.digest.data, y.digest.data)
-  return digestcm
+  # Continue only when sender timestamps are equal 
+  let
+    digestcmp = cmp(x.digest.data, y.digest.data)
+    pstopiccmp = cmp(x.pubsubTopic, y.pubsubTopic)
+
+  if x.senderTime == 0 and
```

Comment:
Just a suggestion, it could have been easier to do an equality check at the beginning of the `cmp` proc  using `==` with an immediate return (if x and y are equal). Then the rest of the logic would become a bit simpler.

Reply:
😆 You're right! Much simpler this way. Done in [d29d0e4](https://github.com/status-im/nim-waku/pull/864/commits/d29d0e44eb3aca5dbfabde343ee8d4e7ac4a1481)